Constructor logic is refactored for the delayed execution of the placeholder refactoring algorithm

d-saravanan opened the pull request:
**Summary**
This pull-request has the logic that will lazy initialize the `_format`  field so that when the `Format`  method is invoked with the arguments, the string is processed and then used for transformation. 
Hence, there will be faster object construction and similar to that of the existing implementation..

Kindly review this and share your review comments. 

Addresses #763

pakrym commented on src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs:
```diff
@@ -18,24 +18,36 @@ public class LogValuesFormatter
         public LogValuesFormatter(string format)
         {
             OriginalFormat = format;
+            _format = new Lazy<string>(() => FormatInput(format), true);
```
2 additional allocations per format string.

Just make FormatInput check the _format field and calculate when it's null.

d-saravanan replied:
Used this lazy loading logic as the variable was readonly. Refactored this code as per the comments and checked-in now.

pakrym commented on src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs:
```diff
@@ -18,24 +18,36 @@ public class LogValuesFormatter
+            _format = new Lazy<string>(() => FormatInput(format), true);
+        }
+
+        public string OriginalFormat { get; private set; }
```
Why?

d-saravanan replied:
Since _format was readonly, used this to lazily load the value

pakrym commented on src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs:
```diff
@@ -18,24 +18,36 @@ public class LogValuesFormatter
+        }
+
+        public string OriginalFormat { get; private set; }
+        public List<string> ValueNames => _valueNames;
```
Why?

d-saravanan replied:
refactored and checked-in. Please review the latest check-in and share review comments.

pakrym commented:
@d-saravanan tests fail.

d-saravanan commented:
@pakrym I have updated the code and fixed the test cases. The Travis-CI build has passed in Linux  and have failed in Mac version due to dotnet build. However, in my repository, the Travis-CI builds are passing. Can you please trigger a build and let me know review comments if any.

pakrym commented on src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs:
```diff
@@ -18,24 +18,46 @@ public class LogValuesFormatter
 
-            while (scanIndex < endIndex)
+            //There should be an open and a closed brace, absence of which, results in unwanted processing.
+            if((!format.Contains(openBrace) && !format.Contains(closingBrace))
```
Previous code was optimized to traverse string only ones in search of `{` or `}`, new code would do it multiple times.

In the case where the string doesn't contain braces `_format` would never be set and string would get re-scanned every time `FormatInput` is called.

d-saravanan replied:
Sorry, the check made only considering the negative case. Since the contains will cause additional search operation, the conditions are removed. 

Also, the _format is set even if the input string does not contain any braces.

Please review the check-in and share your comments.

pakrym commented on src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs:
```diff
@@ -18,24 +18,46 @@ public class LogValuesFormatter
+
+        private string FormatInput(string format)
+        {
+            if(!string.IsNullOrEmpty(_format))
```
Space after if

d-saravanan replied:
code formatting done for the full file.

pakrym commented on src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs:
```diff
@@ -18,24 +18,46 @@ public class LogValuesFormatter
-        private readonly string _format;
-        private readonly List<string> _valueNames = new List<string>();
+        private string _format;
+        private List<string> _valueNames = new List<string>();
```
Do we need to eagerly allocate _valueNames?

d-saravanan replied:
The allocation is modified to be initialized when there are braces in the input format and suitable null checks are added.

d-saravanan commented:
@pakrym I have checked-in the code addressing your review comments. Also, I wrongly started a review thinking that it will raise to your name. Please review the changes and let me know your review comments.

pakrym commented on src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs:
```diff
@@ -17,25 +17,47 @@ public class LogValuesFormatter
+            char openBrace = '{', closingBrace = '}';
 
+            //{} => Nothing to process, min length should be greater than 2. Ex: "{0}"
+            if ((scanIndex == endIndex) || format.Length < 3) 
```
`scanIndex == endIndex` Length==0? It's covered by second condition already.

d-saravanan replied:
removed the condition and checked-in.

pakrym commented on src/Microsoft.Extensions.Logging.Abstractions/Internal/LogValuesFormatter.cs:
```diff
@@ -17,25 +17,47 @@ public class LogValuesFormatter
+
+        private string FormatInput(string format)
+        {
+            if (!string.IsNullOrEmpty(_format))
```
Why isn't null check enough?

d-saravanan replied:
This is initialized with an empty list because we have directly used the variable without null checks in the LoggerMessage.cs, FormattedLogValues.cs & LogFormatAnalyzer.cs. If I set null value here, those files will have to check for null and then add to the KeyValuePair. The references are listed below for your reference. Kindly confirm

![image](https://user-images.githubusercontent.com/20381669/37871832-5110b5fa-3015-11e8-9737-431bc2496d84.png)

d-saravanan commented:
@pakrym can you please let me know if the changes are fine.

d-saravanan commented:
@pakrym I have made the changes and checked-in. Please review and let me know if it is fine.

pakrym commented:
Closing this PR. It hasn't been updated in a long time and tests are failing.
